# Overview
This is a from scratch standard implementation of the Support Vector Machine algorithm, and how it is able to create a non-linear decision boundary for datasets that are mapped by some complex function, with non-separable datapoints for binary classification. We will be using the Kernel trick in conjunction with the SVM algorithm to demonstrate the usefulness of this algorithm, and comparing it with our standard Logistic Regression GLM implementation.

# Theory for SVM + Kernel Trick
Suppose a dataset $\mathcal{D} = \{x^{(i)}, \, y^{(i)}\}_{i = 1}^{m}$ with m samples, wherein $x^{(i)} \in \mathbb{R}^{n}$ is an n-feature vector, and $y^{(i)} \in \{0, \, 1\}$ is a binary output variable. For simplicity of explanation, we put forward the concept of a geometric and functional margin, assuming dataset $\mathcal{D}$ to be separable by a linear decision boundary- or in a higher dimension as we call it, a hyperplane.

## Defining the hyperplane:

A hyperplane $H$ in an $\mathbb{R}^n$ space is a set of points given by the n-feature vector $\vec{x}$, governed by the equation:
$$a^{T} \vec{x} + b = 0$$
where, $a \in \mathbb{R}^n$, $b \in \mathbb{R}$. Here, the phrase "separable datapoints with respect to $H$" imply that if there is a function $f : \mathbb{R}^{n} \longrightarrow \{0, \, 1\}$, then $\vec{x_1} \in H_1$ where $H_1$ is the halfspace set of points governed by the equation $a^{T} \vec{x_1} + b \ge 0$ and $\vec{x_2} \in H_2$ where $H_2$ is the complementary halfspace of $H_1$ given by $a^{T} \vec{x_2} + b < 0$, where $\vec{x_1}, \, \vec{x_2} \in \mathcal{D_x}$, where $\mathcal{D_x}$ is the dataset space of the inputs. Using this, we shall demonstrate the algorithm for SVM + Kernel trick.

## Logistic Regression and SVM's - The similarity:

If we look at the binary classification algorithm from the lens of Logistic Regression GLM, we would expect a well performing predictive parameter $\theta$ to function as the following (assuming we have added the bias term, implying $x^{(i)} \in \mathbb{R}^{n + 1}$) - 
1. If $y^{(i)} = 1$ for the $i^{th}$ sample, then $\theta^{T} x^{(i)} >> 0$ (as this would imply our hypothesis function $g(z^{(i)} = \theta^{T} x^{(i)}) \approx 1$).
2. If $y^{(i)} = 0$, then $\theta^{T} x^{(i)} << 0$ (as this would imply our hypothesis function $g(z^{(i)} = \theta^{T} x^{(i)}) \approx 0$).
In the geometric space, this can be interpreted as the following statement - The greater the euclidean distance of the datapoint from the hyperplane $\theta^{T} x = 0$, the better the model performs, given that the datapoint lies on the correct side of the hyperplane.

## The geometric margin:

Henceforth, we define a new parameter $\gamma^{(i)} = \frac{y^{(i)} (\omega^{T} x^{(i)})}{||\omega||_2}$ for the $i^{th}$ sample, which is going to determine how well our model is performing in predicting the binary output for that sample (We now drop the bias term convention, and have $x^{(i)} \in \mathbb{R}^n$ along with $\theta \in \mathbb{R}^{n + 1}$ splitting into $\omega \in \mathbb{R}^n$ and $b \in \mathbb{R}$).
We can generalize this parameter for the whole dataset, such that $\gamma = \min\limits_i \, \gamma^{(i)}$, denotes the worst-case sample for our model. In geometric space, this is denoted by the datapoint closest to the hyperplane, implying the worst case performance, or the lower bound of the performance of our model. This worst case performance is what we call the geometric margin.

## Our optimization problem:

If we can improve this lower bound to be at a larger value, then we get an overall better performing model. In mathematical terms, we state this problem as-
$$\max\limits_{\gamma, \, \omega, \, b} \gamma \; ; \; \frac{y^{(i)}(\omega^{T} x^{(i)} + b)}{||\omega||_2} \ge \gamma \; \forall \; i$$
This problem can then be transformed into a convex optimization problem, given by-
$$\min\limits_{\omega, \, b} ||\omega||_2 \; ; \; y^{(i)}(\omega^{T} x^{(i)} + b) \ge 1 \; \forall \; i$$
where $y^{(i)}(\omega^{T} x^{(i)} + b)$ is called the functional margin with respect to hyperplane $H$ for the $i^{th}$ sample.
Now using "Representer's theorem", we can represent our parameter $\omega = \sum_{i = 1}^{m} \alpha_i y^{(i)} x^{(i)}$ where then we denote the parameter $\vec{\alpha} \in \mathbb{R}^{m}$ containing each of its entries as $\alpha_i$. $\omega$ is represented as a linear combination of the all our datapoints via representer's theorem, which states we can do so without losing any information. This transforms our optimization problem into-
$$J(\vec{\alpha}) = \max\limits_{b, \, \vec{\alpha}} \sum_{i = 1}^{m} \alpha_i \; - \;\frac{1}{2}\sum_{i = 1}^{m} \sum_{j = 1}^{m} \alpha_i \alpha_j y^{(i)} y^{(j)} ((x^{(i)})^{T} x^{(j)}) \; ; \; \alpha_i \ge 0, \, \sum_{i = 1}^{m} y^{(i)} \alpha_i = 0$$

## The Kernel Trick:
We introduce the concept of "Inner products", denoted by the symbols <>. This is written as $\langle x, z \rangle = x^{T} z$. This particular product will be used to form the Kernel trick for SVM algorithms, and we will also use these to expand our feature space into a larger dimensional feature space, without increasing computation cost.
1. Let's define a function $\phi : \mathbb{R}^n \longrightarrow \mathbb{R}^{n + k}$, where we can arbitrarily set k. This is our transformation function, which allows us to expand the feature space to include finer details of the datapoint features.
2. These extra features will essentially be combinations of the previous input features, implying that the added features to the datapoint input vector are not independent of the features actually available to us. However, we can use these features to predict a non-linear decision boundary.
3. Thus, we replace $x^{(i)} \in \mathbb{R}^n \, \forall i$ with $\phi(x^{(i)}) \in \mathbb{R}^{n + k}$, and use these input vectors now to optimize our decision boundary.
4. This transforms the optimization problem into $\max\limits_{b, \, \vec{\alpha}} J(\vec{\alpha}) = \sum_{i = 1}^{m} \alpha_i \; - \;\frac{1}{2}\sum_{i = 1}^{m} \sum_{j = 1}^{m} \alpha_i \alpha_j y^{(i)} y^{(j)} (\phi((x^{(i)}))^{T} \phi(x^{(j)})) \; ; \; \alpha_i \ge 0, \, \sum_{i = 1}^{m} y^{(i)} \alpha_i = 0$ where we replaced $((x^{(i)})^{T} \, x^{(j)}) = \langle x^{(i)}, \, x^{(j)} \rangle$ with $\langle \phi((x^{(i)})), \, \phi(x^{(j)}) \rangle$, where $\langle \phi((x^{(i)})), \, \phi(x^{(j)}) \rangle$ is called the Kernel function, $K(\langle x^{(i)}, x^{(j)} \rangle)$.
5. Now, the Kernel trick is essentially just a faster way to compute this higher dimensional inner-product. For example, let's assume that we want to higher dimensional feature vector that just combines the input features into all possible combinations. For example- if our feature vector has $n = 2$ features $x_1$ and $x_2$, then we can add a total of $2^n - 1 - n$ more features present as combinations of the previous 2 features- in this case, $x_1 x_2, \; x_1, \; x_2$ are all the possible features of the higher dimensional vector we are aiming to form.
6. Let's say we are aiming to form a Kernel that computes all the monomial terms that can be formed from a given number of features. Then, we use the polynomial kernel function for this, that defines $K(\langle x, z \rangle) = (x^{T}z + c)^d$. This will accomodate all the monomial terms + constant terms inside the higher dimension vector $\phi(x)$ such that $\phi : \mathbb{R}^{n} \longrightarrow \mathbb{R}^{n^2 + n + 1}$.
7. Using this function, we form a Kernel matrix $K_{m \, x \, m}$, such that $K_{ij} = K(\langle x^{(i)}, x^{(j)} \rangle) = ((x^{(i)})^{T} x^{(j)} + c)^d$ is the ij matrix element.
8. Our main optimization problem then reduces into a simple gradient ascent problem in terms of the parameter set $\vec{\alpha}$, with the gradient of the maximizer function $J(\vec{\alpha})$ being expressed in terms of the $K$ matrix, a diagonal matrix $Y_{m \, x \, m}$ containing all entries of Y from the dataset, and a constant column vector $\vec{1}$ with m entries. This will fetch us $$\nabla_{\vec{\alpha}} J(\vec{\alpha}) = \vec{1} - (Y \,K \,Y) \, \vec{\alpha}$$
9. Upon performing gradient ascent on this maximizer function (while respecting the constraints of the problem), we shall find the optimal $\vec{\alpha}^*$ that we can use to form our linear geometric margin in the higher dimension space. The beauty of this algorithm is that this linear geometric margin in the higher dimensional space, when projected onto the actual n-dimension space, will form a non-linear decision boundary. It's a matter of prespective.

# Objective:
This implementation aims to demonstrate the from scratch code and implementation behind SVM's and Kernels, and how they perform on a non-linear curved decision boundary.

In [2]:
#Importing all the basic libraries required.
import matplotlib.pyplot as plt
import numpy as np
import pandas as pd
from sklearn.preprocessing import StandardScaler
import seaborn as sns

In [3]:
#Generating random seed for reproducibility of results from code. 99 is an arbitrary choice.
np.random.seed(99)
#We create an input matrix X with values uniformly distributed between [-500, +500].
#This is a 5-feature matrix, with 500 training samples.
X_raw = np.random.uniform(low = -500, high = +500, size = (500, 5))
#We define our own non-linear non-trivial function f in terms of the features of input vector x, to simulate a non-linear decision boundary.
f = ((X_raw[:, 0] ** 2) + (2 * X_raw[:, 0] * X_raw[:, 1]) - (X_raw[:, 1] ** 2) + (2 * X_raw[:, 1] * X_raw[:, 2]) - (X_raw[:, 0] * X_raw[:, 1] * X_raw[:, 2] * X_raw[:, 3] * X_raw[:, 4]) + 47)
#We put the signum function on this f, to make a binary encoded output and store it in a column vector Y.
Y_raw = (f >= 0).astype(int).reshape(-1, 1)
#Verifying the shapes.
print(X_raw.shape)
print(Y_raw.shape)

(500, 5)
(500, 1)


In [4]:
#Split the X matrix to form a training set and a queryset. Same for Y.
#In our implementation, we do not split theta into omega and b. Hence, we also need to add the biasing term.
X_dataset = X_raw[:400, :]
X_dataset = np.hstack([np.ones((400, 1)), X_dataset])
Y_dataset = Y_raw[:400, :]
X_queryset = X_raw[-100:, :]
X_queryset = np.hstack([np.ones((100, 1)), X_queryset])
Y_true_op = Y_raw[-100:, :]
#Verifying the shapes.
print(X_dataset.shape)
print(Y_dataset.shape)
print(X_queryset.shape)
print(Y_true_op.shape)

(400, 6)
(400, 1)
(100, 6)
(100, 1)


In [6]:
#Processing the data and scaling it to numerically stable values (prevents overflow).
scaler = StandardScaler()
X = scaler.fit_transform(X_dataset)
#Defining the gradient function for logistic regression.
def grad(X_inp, Y_inp, Q_inp):
    Z = X_inp @ Q_inp
    G = 1/(1 + np.exp(-Z))
    return X_inp.T @ (Y_inp - G)
#Defining the logistic regression function to run on our training set, and use that function to predict outputs for our queryset.
def logistic(X_inp, Y_inp, X_queries):
    Q1 = np.zeros((X_inp.shape[1], 1))
    i = 0
    iters = 5000
    alpha = 0.05
    #We perform gradient ascent on the theta for logistic GLM, to get our main parameter vector theta.
    #5000 iterations and 0.05 learning rate is a design choice.
    while(i < 5000):
        Q1 += alpha * grad(X_inp, Y_inp, Q1)
        i = i + 1
    Z = X_queries @ Q1
    #Clipping the output of z to prevent overflow or underflow of values when fed into the exponential function in the sigmoid output.
    Z_clipped = np.clip(Z, -500, +500)
    Y_predicted = 1/(1 + np.exp(-Z_clipped))
    return Y_predicted
#Running the function.
logistic_regression = logistic(X, Y_dataset, X_queryset)
#Verifying the shapes.
print(X.shape)
print(Y_dataset.shape)
print(logistic_regression.shape)

(400, 6)
(400, 1)
(100, 1)


In [11]:
#Defining the SVM function for SVM + Kernel classifier implementation.
def svm(X_inp, Y_inp, X_queries, gamma, degree):
    Y_in = Y_inp
    Y_in = 2 * Y_in.flatten() - 1
    #Our real Y_in will have 0's and 1's as entries. SVM assumes -1 and +1 as the binary output entries.
    #Defining the Y as a diagonal matrix, and K as the Kernel matrix for our SVM implementation.
    #Since our function f contains monomial terms, it is best to use the polynomial kernel to map our input vector to the higher dimension feature space.
    Y_d = np.diag(Y_in)
    K = (gamma * (X_inp @ X_inp.T) + 1) ** degree
    #Initializing the alpha parameter vector at the origin. This vector implicitly contains the bias term as well.
    #Defining the gradient using this alpha vector, K and Y_d.
    A = np.zeros((X_inp.shape[0], 1))
    Q2 = K * (Y_in[:, None] * Y_in[None, :])
    #5000 iterations is a design choice.
    i = 0
    iters = 5000
    while(i < 5000):
        gradient = np.ones((X_inp.shape[0], 1)) - (Q2 @ A)
        A += 0.05 * gradient
        i += 1
        #We respect the constraints, and upon each iteration we clip the alpha vector between 0 and 1e6 (this is the constant C as the upper bound in our constraints).
        #In actuality, doing so will give us a sub-optimal solution, since it might so happen that the alpha gets stuck on to the origin and never reach the optimal value.
        A = np.clip(A, 0, 1e6)
        #Also, since the vector alpha should lie in the hyperplane defined by alpha.T @ Y(column vector) = 0, we project our alpha onto that hyperplane so as to respect the constraints, again.
        #This method was suggested by chatGPT, and ensured actual accuracy in the algorithm.
        A -= ((np.dot(Y_in, A) / np.dot(Y_in, Y_in)) * Y_in).reshape(-1, 1)
    #Forming the kernel matrix for the query space.
    #Returning predicted values.
    K_queries = (gamma * (X_queries @ X_inp.T) + 1) ** degree
    hyperplane = K_queries @ (A.flatten() * Y_in)
    Y_predicted = (hyperplane > 0).astype(int).reshape(-1, 1)
    return Y_predicted.reshape(-1, 1)
#Gamma and degree are arbitrary hyperparameters that we have used for our polynomial kernel.
svm_predicted = svm(X, Y_dataset, X_queryset, 1e6, 5)
#Verifying shapes.
print(svm_predicted.shape)

(100, 1)


In [12]:
#Comparing accuracies of our SVM + Kernel algorithm and our logistic regression algorithm on our non-linear decision boundary.
svm_correct = (svm_predicted - Y_true_op == 0).astype(int)
svm_acc = np.sum(svm_correct, axis = 0)
logistic_preds = (logistic_regression > 0.5).astype(int)
logistic_correct = (logistic_preds - Y_true_op == 0).astype(int)
logistic_acc = np.sum(logistic_correct, axis = 0)
print("Accuracy with logistic regression:", logistic_acc[0] / 100)
print("Accuracy with SVM method:", svm_acc[0] / 100)

Accuracy with logistic regression: 0.46
Accuracy with SVM method: 0.62


In [13]:
#Verifying our SVM implementation with the help of a pre-made Kernel pipeline.
from sklearn.svm import SVC
clf = SVC(kernel = 'poly', degree = 5, C = 1e6, gamma = 1e6, coef0 = 1)
clf.fit(X, Y_dataset.flatten())
Y_pred = clf.predict(X_queryset)
from sklearn.metrics import accuracy_score
Y_pred = clf.predict(X_queryset)
accuracy = accuracy_score(Y_true_op.flatten(), Y_pred)
print("Accuracy:", accuracy)

Accuracy: 0.84


Clearly, our SVM + polynomial kernel method has performed much better as a classifier for this non-linear decision boundary dataset. 

## Some problems I faced, and what I learned:
1. At first, I thought that the closed form analytical solution by putting gradient of J as 0 would give me the same result as the one solved respecting the constraints. I spent some time figuring out the closed form analytical solution for that, only to yield a low accuracy of 42%.
2. Used the clipping method to take into account the $0 \le \alpha_i \le C$ and improved the accuracy to a whopping 45% (sarcasm).
3. I then figured out I had not simulated a proper function for defining the binary output dataset, and had in fact simulated those values randomly. Also, since I was using the analytical solution at that time, I had to reduce my sample space from 5000 to 500 as the computation cost for calculating the inverse (a part of my closed form solution) was heavy. Did all of that only to yield a decreased accuracy of 43% (wow).
4. Found that I had been using the wrong method for Kernel implementation. If we have a training set $\mathcal{D}$ and a queryset $\mathcal{Q}$, then we have to run 2 separate Kernel matrices for each of them. The training set Kernel matrix helps in calculating the parameter vector $\vec{\alpha}$, which is then carried on to the queryset. However, we then calculate another queryset Kernel matrix $K_{\mathcal{Q}}$, which has $K_{ij} = [p  (x^{(i)}_{\mathcal{Q}})^{T} x^{(j)}_{\mathcal{D}} + q]^{d}$ as the element of $i^{th}$ row and $j^{th}$ column, where $x_{\mathcal{Q}} \in \mathcal{Q}$ and $x_{\mathcal{D}} \in \mathcal{D}$ belongs to the training dataset. This boosted my accuracy to a hard 49%.
5. I wanted to almost give up, but then I realised that putting the analytical solution as the true value of $\vec{\alpha}$ ignores the constraints completely, since the analytical solution is calculated over the whole hyperspace of $R^{m}$. This will obviously give a below satisfactory value.
6. Implemented a gradient ascent method instead, with clipping the values of $\alpha_i$ from $[0, \, 10^6]$ upon each iteration, along with projecting the steps onto the $(\vec{\alpha})^T Y = 0$ hyperplane, where $Y$ denoted the column matrix for the output binaries of the training set here. The projection idea was given to me by chatGPT, since I was not able to think of how we would exactly implement the second constraint. Achieved an actually decent accuracy of 62% after the whole mess.

This was a fruitful session, and a LONG one at that (8.6 hours, to be precise). Anyway, thank you for sticking around! :)